### PR TITLE
Implement google cloud storage backend

### PR DESCRIPTION
**Pull request by mjgarton**
This is the initial implementation of a gcs backend.

### CODE REVIEW

File: straw_gcs.go [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+)
+
+var _ StreamStore = &GCSStreamStore{}
+
+func NewGCSStreamStore(credentialsFile string, bucket string, options ...S3Option) (*GCSStreamStore, error) {
```
Comment (nylar) [inline straw_gcs.go]: `S3Option`?

Comment (mjgarton) [reply straw_gcs.go]: doh. fixed.

File: straw_gcs.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+			return nil, os.ErrNotExist
+		}
+		return nil, err
+	}
+
+	return &GCSReader{r, fs, nameNoSlash}, nil
+}
+
+type GCSReader struct {
+	rc      *storage.Reader
```
Comment (johanbrandhorst) [inline straw_gcs.go]: If this was embedded instead,  `Read` and `Close` are promoted implicitly. I quite like them being explicit, but maybe something to consider?

File: straw_gcs.go [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+	}
+
+	if _, err := fs.Stat(name); err == nil {
+		return fmt.Errorf("%s : file exists", name)
+	}
+
+	obj := fs.client.Bucket(fs.bucket).Object(name)
+	w := obj.NewWriter(context.Background())
+
+	if _, err := w.Write([]byte{}); err != nil {
```
Comment (johanbrandhorst) [inline straw_gcs.go]: Is this really how you create a directory in GCS?

File: straw_gcs.go [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+	if wantSlash {
+		if !strings.HasSuffix(s, "/") {
+			return s + "/"
+		}
+	} else {
+		if strings.HasSuffix(s, "/") {
+			return s[0 : len(s)-1]
+		}
+	}
+	return s
```
Comment (johanbrandhorst) [inline straw_gcs.go]:
```go
if !wantSlash {
	return strings.TrimSuffix(s, "/")
}
if !strings.HasSuffix(s, "/") {
	return s + "/"
}
return s
```
is a little simpler.

File: straw_gcs.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+}
+
+func (fs *GCSStreamStore) lastElem(s string) string {
+	_, f := filepath.Split(fs.noSlashSuffix(s))
+	return f
+}
+
+type gcsUploader struct {
+	wc io.WriteCloser
```
Comment (johanbrandhorst) [inline straw_gcs.go]: Ditto embedding

File: straw_gcs.go (hunk 3 above)
Comment (mjgarton) [reply straw_gcs.go]: I didn't find any conclusive spec of the "proper" way, but this works has been used somewhere else (probably wherever I found it on stack overflow or something)

If there's a better way that definitely works, I'd be happy to change.

File: straw_gcs.go (hunk 5 above)
Comment (mjgarton) [reply straw_gcs.go]: Actually, I don't need to embed this at all. Will fix.

File: straw_gcs.go (hunk 2 above)
Comment (mjgarton) [reply straw_gcs.go]: fixed.

File: straw_gcs.go (hunk 4 above)
Comment (mjgarton) [reply straw_gcs.go]: done.

**Pull request merged**